src/sql/workbench/services/connection/browser/connectionManagementService.ts

```diff
@@ -874,6 +876,61 @@ export class ConnectionManagementService extends Disposable implements IConnecti
 		return false;
 	}
 
+	/**
+	 * Refresh Azure access token if it's expired.
+	 * @param uri connection uri
+	 * @returns true if no need to refresh or successfully refreshed token
+	 */
+	public async refreshAzureAccountTokenIfNecessary(uri: string): Promise<boolean> {
+		let profile = this._connectionStatusManager.getConnectionProfile(uri);
+		if (!profile) {
+			this._logService.warn(`Connection not found for uri ${uri}`);
+			return false;
+		}
+
+		//wait for the pending reconnction promise if any
+		let previousReconnectPromise = this._uriToReconnectPromiseMap[uri];
+		if (previousReconnectPromise) {
+			this._logService.info(`Found pending reconnect promise for uri ${uri}, waiting.`);
+			try {
+				let previousConnectionResult = await previousReconnectPromise;
+				if (previousConnectionResult && previousConnectionResult.connected) {
+					this._logService.info(`Previous pending reconnection for uri ${uri} succeeded.`);
+					return true;
+				}
+				this._logService.info(`Previous pending reconnection for uri ${uri} failed with error ${previousConnectionResult.errorMessage}, error code: ${previousConnectionResult.errorCode}.`);
+			} catch (err) {
+				this._logService.info(`Previous pending reconnect promise for uri ${uri} is rejected with error ${err}, will attempt to reconnect if necessary.`);
+			}
+		}
+
+		let expiry = profile.options.expiresOn;
+		if (typeof expiry === 'number' && !Number.isNaN(expiry)) {
+			const currentTime = new Date().getTime() / 1000;
+			const maxTolerance = 2 * 60; // two minutes
```

Comment:
where did two minutes come from?  is this a recommendation from somewhere or just a reasonable guess?

Reply:
This comes from the existing two min tolerance window from https://github.com/microsoft/azuredatastudio/blob/ccaf111696aac9c246b9ec4f969cb6a8d4f5b39e/extensions/azurecore/src/account-provider/auths/azureAuth.ts#L188
It was set to 5min before the current 2min version.
https://github.com/microsoft/azuredatastudio/blob/5613a97fae6b5cb37f079a67c33c4694f0046c17/extensions/azurecore/src/account-provider/auths/azureAuth.ts#L211

I don't think this comes from any recommendation, more like a guess. So I did the same tolerance window for token refresh check.